src/audit/bypassable_contains_conditions.rs

```diff
@@ -0,0 +1,159 @@
+use github_actions_models::common::{expr::ExplicitExpr, If};
+
+use super::{audit_meta, Audit};
+use crate::{
+    expr::{Context, Expr},
+    finding::{Confidence, Severity},
+    models::JobExt,
+};
+
+const USER_CONTROLLABLE_CONTEXTS: [&str; 15] = [
+    "env.GITHUB_ACTOR",
+    "env.GITHUB_BASE_REF",
+    "env.GITHUB_HEAD_REF",
+    "env.GITHUB_REF",
+    "env.GITHUB_REF_NAME",
+    "env.GITHUB_SHA",
+    "env.GITHUB_TRIGGERING_ACTOR",
+    "github.actor",
+    "github.base_ref",
+    "github.head_ref",
+    "github.ref",
+    "github.ref_name",
+    "github.sha",
+    "github.triggering_actor",
+    "inputs.",
+];
+
+pub(crate) struct BypassableContainsConditions;
+
+audit_meta!(
+    BypassableContainsConditions,
+    "bypassable-contains-conditions",
+    "bypassable contains conditions checks"
+);
+
+impl Audit for BypassableContainsConditions {
+    fn new(_state: super::AuditState) -> anyhow::Result<Self>
+    where
+        Self: Sized,
+    {
+        Ok(Self)
+    }
+
+    fn audit_normal_job<'w>(
+        &self,
+        job: &super::NormalJob<'w>,
+    ) -> anyhow::Result<Vec<super::Finding<'w>>> {
+        let conditions = job
+            .r#if
+            .iter()
+            .map(|cond| (cond, job.location()))
+            .chain(
+                job.steps()
+                    .filter_map(|step| step.r#if.as_ref().map(|cond| (cond, step.location()))),
+            )
+            .filter_map(|(cond, loc)| {
+                if let If::Expr(expr) = cond {
+                    Some((expr.as_str(), loc))
+                } else {
+                    None
+                }
+            });
+
+        conditions
+            .flat_map(|(expr, loc)| {
+                Self::insecure_contains(expr).into_iter().map(move |(confidence, context)| {
+                    Self::finding()
+                        .severity(Severity::High)
+                        .confidence(confidence)
+                        .add_location(
+                            loc.with_keys(&["if".into()])
+                                .primary()
+                                .annotated(format!("contains(..) condition can be bypassed if attacker can control '{context}'")),
+                        )
+                        .build(job.parent())
+                })
+            })
+            .collect()
+    }
+}
+
+impl BypassableContainsConditions {
+    fn walk_tree_for_insecure_contains<'a>(
+        expr: &'a Expr,
+    ) -> Box<dyn Iterator<Item = (&'a str, &'a Context<'a>)> + 'a> {
+        match expr {
+            Expr::Call {
+                func: "contains",
```

Comment:
NB: functions are case-insensitive in GHA expressions, so this should probably be `func.eq_ignore_ascii_case("contains")`